Sources/SwiftProcessing/Core/AudioIn.swift

```diff
@@ -0,0 +1,237 @@
+    private let settings: [String:Any] = [
+        AVFormatIDKey: kAudioFormatLinearPCM,
+        AVSampleRateKey: 44100,
+        AVNumberOfChannelsKey: 1,
+        AVLinearPCMBitDepthKey: 16,
+        AVLinearPCMIsBigEndianKey: false,
+        AVLinearPCMIsFloatKey: false
+    ]
+    
+    // Change the multiplier to amplify the effect of sound on your shapes.
+    // You can, optionally, create more than one multiplier.
+    
+    /// A multiplier that enables you to magnify the effect of the microphone input.
+    /// Larger = bigger.
+    
+    public var multiplier = 1.0
+    
+    /// The property where the audio level is stored. This includes the effect of any mulitplier.
+    
+    public var level: Double!
+    
+    /*
+     * MARK: - INIT
+     */
+    
+    /// AudioIn() is a singleton object that is associated with the microphone.
+    /// Singleton's ensure that only one object is created. They're often used when
+    /// they are associated with a single piece of hardware, like a camera or a
+    /// microphone. You can refer direclty to AudioIn().shared to access audio input.
+    
+    public static let shared = AudioIn()
+    
+    private init() {
+        print("Creating audio in object")
+        // Set up the microphone to start listening.
+        
+        // Original code. Replaced with Apple's suggested workflow.
+        
+        /*
+        let audioSession = AVAudioSession.sharedInstance()
+        do {
+            try audioSession.setCategory(AVAudioSession.Category.playAndRecord)
+        } catch let error as NSError {
+            print(error.description)
+        }
+         */
+
+        setupAudioSession()
+        enableBuiltInMic()
+        
+        do {
+            let url = URL(string: NSTemporaryDirectory().appending("tmp.caf"))!
+            print("recording to")
+            print(url)
+            try recorder = AVAudioRecorder(url: url, settings: settings)
+        } catch {
+            print("error!")
+        }
+        
+        switch AVAudioSession.sharedInstance().recordPermission {
+        case AVAudioSession.RecordPermission.granted:
+            print("Permission granted")
+        case AVAudioSession.RecordPermission.denied:
+            print("Pemission denied")
+        case AVAudioSession.RecordPermission.undetermined:
+            print("Request permission here")
+            AVAudioSession.sharedInstance().requestRecordPermission({ (granted) in
+                // Handle granted
+            })
+        }
+    }
+    
+    // Source: https://developer.apple.com/documentation/avfaudio/avaudiosession/capturing_stereo_audio_from_built-in_microphones
+    
+    func setupAudioSession() {
+        do {
+            let session = AVAudioSession.sharedInstance()
+            try session.setCategory(.playAndRecord, options: [.defaultToSpeaker, .allowBluetooth])
+            try session.setActive(true)
+        } catch {
+            fatalError("Failed to configure and activate session.")
```

Comment:
We should avoid fatal errors as it'd be hard to debug for a newcomer. Why would this ever fail? Could we provide a clue to the user?

Something we could explore is a user friendly error screen that takes over the view, similar to Apple's yellow screen with a red cross out,  seen here https://stackoverflow.com/questions/63135404/why-do-some-views-appear-as-a-red-no-entry-sign-in-widgets . It could be the sketch version of a fatal error

Reply:
I'll include this change in the next pull request.